File: apps/worker/src/worker/openai_images.py
import base64
import binascii
from io import BytesIO

from openai import APIConnectionError, APIStatusError, APITimeoutError, OpenAI
from openai import OpenAIError as OpenAISDKError

from .validation import DEFAULT_OPENAI_BASE_URL, DrawTaskInput


class ImageGenerationError(Exception):
    def __init__(self, message: str, status_code: int = 502) -> None:
        super().__init__(message)
        self.status_code = status_code


def generate_openai_image(
    task_input: DrawTaskInput,
    api_key: str,
    base_url: str | None,
    source_image_bytes: bytes | None = None,
) -> list[bytes]:
    if not api_key:
        raise ImageGenerationError("缺少 OpenAI API Key，请先在页面左侧配置。", 401)

    try:
        client = OpenAI(
            api_key=api_key,
            base_url=resolve_base_url(base_url),
            timeout=110,
        )
        if source_image_bytes is None:
            response = client.images.generate(
                model=task_input.model,
                prompt=task_input.prompt,
                n=task_input.output_count,
                size=task_input.size,
                quality=task_input.quality,
                output_format="png",
                response_format="b64_json",
            )
        else:
            response = client.images.edit(
                model=task_input.model,
                image=to_uploadable_image(source_image_bytes),
                prompt=task_input.prompt,
                n=task_input.output_count,
                size=task_input.size,
                quality=task_input.quality,
                output_format="png",
                response_format="b64_json",
            )
    except APIStatusError as error:
        raise ImageGenerationError(
            extract_openai_error_message(error),
            error.status_code,
        ) from error
    except APITimeoutError as error:
        raise ImageGenerationError("OpenAI 图像生成超时，请稍后重试。") from error
    except APIConnectionError as error:
        raise ImageGenerationError(f"调用 OpenAI 图像接口失败：{error}") from error
    except OpenAISDKError as error:
        raise ImageGenerationError(f"调用 OpenAI 图像接口失败：{error}") from error

    if not response.data:
        raise ImageGenerationError("OpenAI 响应中没有图像数据。")

    images: list[bytes] = []

    for item in response.data:
        image = item.b64_json

        if not image:
            raise ImageGenerationError("OpenAI 响应中没有图像数据。")

        try:
            image_bytes = base64.b64decode(image, validate=True)
        except binascii.Error as error:
            raise ImageGenerationError("OpenAI 响应中的图像数据不是有效 base64。") from error

        if not image_bytes.startswith(b"\x89PNG\r\n\x1a\n"):
            raise ImageGenerationError("OpenAI 响应中的图像不是 PNG 格式。")

        images.append(image_bytes)

    return images


def to_uploadable_image(image_bytes: bytes) -> tuple[str, BytesIO, str]:
    if not image_bytes.startswith(b"\x89PNG\r\n\x1a\n"):
        raise ImageGenerationError("来源图片不是有效 PNG，暂不支持继续生成。", 400)

    return ("source.png", BytesIO(image_bytes), "image/png")


def resolve_base_url(base_url: str | None) -> str:
    normalized_base_url = (base_url or DEFAULT_OPENAI_BASE_URL).strip().rstrip("/")

    if not normalized_base_url.startswith(("http://", "https://")):
        raise ImageGenerationError("Base URL 无效，仅支持 http 或 https 地址。", 400)

    return normalized_base_url


def extract_openai_error_message(error: APIStatusError) -> str:
    try:
        response_payload = error.response.json()
    except ValueError:
        return f"OpenAI 接口返回 {error.status_code}。"

    if isinstance(response_payload, dict):
        error_payload = response_payload.get("error")

        if isinstance(error_payload, dict):
            message = error_payload.get("message")

            if isinstance(message, str) and message:
                return message

    return f"OpenAI 接口返回 {error.status_code}。"

File: apps/web/services/ai-canvas/openai-images.ts
import type { BranchMode } from "@/components/domain/branch-mode"
import {
  isValidHttpBaseUrl,
  normalizeBaseUrl,
  type StandaloneImageInput,
} from "@/lib/validations/ai-canvas"

import { apiError, type ApiError } from "./errors"
import { getPngSize } from "./image-storage"
import type { ProviderConfigRecord } from "./types"

const BRANCH_MODE_INSTRUCTIONS: Record<BranchMode, string> = {
  evolve:
    "请基于参考图继续延展，保留整体视觉气质、材质语言与系列感，允许调整构图、镜头、细节与陪体，但不要无关地替换主题。",
  preserve:
    "请严格围绕参考图中的主体进行修改，尽量保持主体身份、核心轮廓、关键造型与辨识特征，仅按提示词做局部调整、补充或润饰。",
  transform:
    "请把参考图作为创作起点进行大幅改造，可以重构场景、风格、色彩与表现手法，但仍应与参考图存在可追溯的视觉关联。",
}

function compileBranchPrompt(prompt: string, branchMode: BranchMode | null) {
  if (!branchMode) return prompt.trim()
  return `${BRANCH_MODE_INSTRUCTIONS[branchMode]}\n\n用户目标：${prompt.trim()}`
}

export async function generateOpenAiImages(
  input: StandaloneImageInput & {
    outputCount: number
    branchMode?: BranchMode | null
    parentAssetId?: string | null
  },
  providerConfig: ProviderConfigRecord,
  sourceImageBytes?: Uint8Array,
) {
  if (!providerConfig.apiKey) {
    throw apiError("缺少 OpenAI API Key，请先在页面左侧配置。", 401)
  }

  const baseUrl = normalizeBaseUrl(providerConfig.baseUrl)
  if (!isValidHttpBaseUrl(baseUrl)) {
    throw apiError("Base URL 无效，仅支持 http 或 https 地址。", 400)
  }

  const endpoint = new URL(
    sourceImageBytes ? "images/edits" : "images/generations",
    `${baseUrl}/`,
  )
  const prompt = compileBranchPrompt(input.prompt, input.branchMode ?? null)

  const response = sourceImageBytes
    ? await postEditRequest(endpoint, providerConfig.apiKey, {
        model: input.model,
        prompt,
        n: input.outputCount,
        size: input.size,
        quality: input.quality,
        image: sourceImageBytes,
      })
    : await fetch(endpoint, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${providerConfig.apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          model: input.model,
          prompt,
          n: input.outputCount,
          size: input.size,
          quality: input.quality,
          output_format: "png",
          response_format: "b64_json",
        }),
      })

  if (!response.ok) {
    throw await parseOpenAiError(response)
  }

  const payload = (await response.json()) as {
    data?: Array<{ b64_json?: string | null }>
  }

  if (!Array.isArray(payload.data) || payload.data.length === 0) {
    throw apiError("OpenAI 响应中没有图像数据。", 502)
  }

  return payload.data.map((item) => {
    if (!item.b64_json) {
      throw apiError("OpenAI 响应中没有图像数据。", 502)
    }

    const imageBytes = Uint8Array.from(Buffer.from(item.b64_json, "base64"))
    getPngSize(imageBytes)
    return imageBytes
  })
}

async function postEditRequest(
  endpoint: URL,
  apiKey: string,
  input: {
    model: string
    prompt: string
    n: number
    size: string
    quality: string
    image: Uint8Array
  },
) {
  const formData = new FormData()
  formData.set("model", input.model)
  formData.set("prompt", input.prompt)
  formData.set("n", String(input.n))
  formData.set("size", input.size)
  formData.set("quality", input.quality)
  formData.set("output_format", "png")
  formData.set("response_format", "b64_json")
  formData.set(
    "image",
    new File([Buffer.from(input.image)], "source.png", { type: "image/png" }),
  )

  return fetch(endpoint, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
    },
    body: formData,
  })
}

async function parseOpenAiError(response: Response): Promise<ApiError> {
  try {
    const payload = (await response.json()) as { error?: { message?: string } }
    const message = payload.error?.message
    if (typeof message === "string" && message) {
      return apiError(message, response.status)
    }
  } catch {
    // Fall through to generic HTTP status message.
  }

  return apiError(`OpenAI 接口返回 ${response.status}。`, response.status)
}
